src/proxy_handler.cpp

```diff
@@ -0,0 +1,206 @@
+
+    while (conn_error && endpoint_iterator != end) {
+      socket.close();
+      socket.connect(*endpoint_iterator++, conn_error);
+    }
+
+    // Try each endpoint until we successfully establish a connection.
+
+    // Form the rrequest. We specify the "Connection: close" header so that the
+    // server will close the socket after transmitting the response. This will
+    // allow us to treat all data up until the EOF as the content.
+    boost::asio::streambuf request_;
+    std::ostream request_stream(&request_);
+    std::string redirect_uri;
+    std::string orig_uri = request.uri();
+    if (uri_prefix_.size() == 1) {
+      redirect_uri = orig_uri;
+    }
+    else {
+      if (uri_prefix_.size() + 1 >= orig_uri.size())
+        redirect_uri = "/";
+      else 
+        redirect_uri = orig_uri.substr(uri_prefix_.size());
+    }
+
+    if (!redirect_uri_.empty()) {
+      redirect_uri = redirect_uri_;
+    }
+    //TODO: Change to "/"
+    request_stream << "GET " + redirect_uri + " HTTP/1.0\r\n";
+    request_stream << "Host: " << proxy_host_ << "\r\n";
+    request_stream << "Accept: */*\r\n";
+    request_stream << "Connection: close\r\n\r\n";
+
+    // Send the request.
+    boost::asio::write(socket, request_);
+
+    // Read the response status line. The response streambuf will automatically
+    // grow to accommodate the entire line. The growth may be limited by passing
+    // a maximum size to the streambuf constructor.
+    boost::asio::streambuf response_;
+    boost::asio::read_until(socket, response_, "\r\n");
+
+    // Check that response is OK.
+    std::istream response_stream(&response_);
+    std::string http_version;
+    response_stream >> http_version;
+    unsigned int status_code;
+    response_stream >> status_code;
+    std::string status_message;
+    std::getline(response_stream, status_message);
+    if (!response_stream || http_version.substr(0, 5) != "HTTP/") {
+      std::cout << "Invalid response\n";
+      return RequestHandler::INVALID_CONFIG;
+    }
+    if (status_code != 200 && status_code != 302) {
+      std::cout << "Response returned with status code " << status_code << "\n";
+      return RequestHandler::INVALID_CONFIG;
+    } else {
+      response->SetStatus(Response::ResponseCode::OK);
+    }
+
+    // Read the response headers, which are terminated by a blank line.
+    boost::asio::read_until(socket, response_, "\r\n\r\n");
+
+    std::string header;
+    size_t index;
+    size_t cr;
+    std::string new_host = "";
+    std::vector<std::pair<std::string, std::string>> header_vector;
+    while (std::getline(response_stream, header) && header != "\r") {
+      index = header.find(':', 0);
+      cr = header.find('\r', index);
+      if (index != std::string::npos) {
+        if (header.substr(0, index).compare("Location") || status_code != 302) {
+        header_vector.push_back(std::make_pair(header.substr(0, index), header.substr(index+2, cr - (index + 2))));
```

Comment:
Split this line up into 2+ lines. Spaces must go before and after arithmetic operators (index + 2, not index+2)